internal/incident/incident.go

```diff
@@ -320,61 +426,89 @@ func (i *Incident) evaluateEscalations(ctx context.Context, tx *sqlx.Tx, ev *eve
 				if escalation.Condition == nil {
 					matched = true
 				} else {
-					cond := &rule.EscalationFilter{
-						IncidentAge:      time.Since(i.StartedAt),
-						IncidentSeverity: i.Severity,
-					}
-
 					var err error
-					matched, err = escalation.Condition.Eval(cond)
+					matched, err = escalation.Condition.Eval(filterContext)
 					if err != nil {
 						i.logger.Warnw(
 							"Failed to evaluate escalation condition", zap.String("rule", r.Name),
 							zap.String("escalation", escalation.DisplayName()), zap.Error(err),
 						)
 
 						matched = false
+					} else if !matched {
+						incidentAgeFilter := filterContext.ReevaluateAfter(escalation.Condition)
+						retryAfter = min(retryAfter, incidentAgeFilter)
 					}
 				}
 
 				if matched {
-					state := &EscalationState{RuleEscalationID: escalation.ID, TriggeredAt: types.UnixMilli(time.Now())}
-					i.EscalationState[escalation.ID] = state
+					escalations = append(escalations, escalation)
+				}
+			}
+		}
+	}
 
-					i.logger.Infof("Rule %q reached escalation %q", r.Name, escalation.DisplayName())
+	if retryAfter != rule.RetryNever {
+		// The retryAfter duration is relative to the incident duration represented by the escalation filter,
+		// i.e. if an incident is 15m old and an escalation rule evaluates incident_age>=1h the retryAfter would
+		// contain 45m (1h - incident age (15m)). Therefore, we have to use the event time instead of the incident
+		// start time here.
+		nextEvalAt := eventTime.Add(retryAfter)
 
-					if err := i.AddEscalationTriggered(ctx, tx, state); err != nil {
-						i.logger.Errorw(
-							"Failed to upsert escalation state", zap.String("rule", r.Name),
-							zap.String("escalation", escalation.DisplayName()), zap.Error(err),
-						)
+		i.logger.Infow("Scheduling escalation reevaluation", zap.Duration("after", retryAfter), zap.Time("at", nextEvalAt))
+		i.timer = time.AfterFunc(retryAfter, func() {
+			i.logger.Info("Reevaluating escalations")
 
-						return errors.New("failed to upsert escalation state")
-					}
+			i.RetriggerEscalations(&event.Event{
+				Type:    event.TypeInternal,
+				Time:    nextEvalAt,
+				Message: fmt.Sprintf("Incident reached age %v", retryAfter),
```

Comment:
This message is now incorrect with the changed value: If you have escalations `incident_age>=1m` and `incident_age>=2m`, the second escalation will also use "age 1m" because that's the difference to the last evaluation.